### PR TITLE
Added guards to avoid segmentation faults when _fit is called with None parameters

### LINKED ISSUE #804: Python: Segmentation fault when `y` or `eval_set` parameters are None
Problem: Segmentation fault when 
catboost version: 0.14.2
Operating System: Mac OSX Sierra 10.12.6 and Linux 3.10.0-693.17.1.el7.
CPU: x86_64

Example code that makes the failure:
```python

import catboost as cb
import pandas as pd
import numpy as np


data = pd.DataFrame({
              'color': np.random.randint(1, 10, 100),
              'article_id': np.random.randint(100, 125, 100),
              'weight': np.random.rand(100),
              'y': np.arange(100) ** 2
})
y = data['y']
data = data.drop('y', axis=1)

cr = cb.CatBoostRegressor(loss_function='RMSE', eval_metric="RMSE")
fit_params={'use_best_model': True, 'cat_features': [0, 1], 'verbose': False}

# Segmentation fault here
cr.fit(data, y, eval_set=(None, None), **fit_params)

# Segmentation fault here as well
cr.fit(None, y, **fit_params)
```
Looking at the code, it seems that some guards are missing. When `y` is None there's an error but the message is cryptic.  Having not `None` paramters might be obvious when using for experiments but when is used inside a software this might happen and having a clear error is better than a seg fault (or a weird message). 

I will submit a small PR for solving this.

### PR DESCRIPTION
**Pull request by mfcabrera**
Fixes #804 

I don't see tests for the python package itself. It would be a good idea to have some for the python wrapper as well, to test the API behaviour. I tested my changes manually making sure a `CatBoostError` exception is raised when certain parameters (IMHO some exceptions here should be ValueError, to be more specific). I also ran the tests using  `ya make -t -A`. 

I hereby agree to the terms of the CLA available at:  https://yandex.ru/legal/cla/?lang=en

### CODE REVIEW

**Comment by david-waterworth** [comment]
@mfcabrera does this cover the case where X is a Pool(), in which case y should be None?

**Comment by mfcabrera** [comment]
@david-waterworth I've changed the condition. I wasn't aware of it. Thanks :)

**Comment by kizill** [comment]
@arcadia-devtools Ship it!

**Comment by arcadia-devtools** [comment]
CLA already signed

**Comment by arcadia-devtools** [comment]
@kizill, internal review request created: [808715](https://a.yandex-team.ru/review/808715)

**Comment by arcadia-devtools** [comment]
Check [808715](https://a.yandex-team.ru/review/808715) failed: 4 test(s) broken

File: catboost/python-package/catboost/core.py [hunk 1]
```diff
@@ -1108,6 +1107,12 @@ def _fit(self, X, y, cat_features, pairs, sample_weight, group_id, group_weight,
              column_description, verbose_eval, metric_period, silent, early_stopping_rounds,
              save_snapshot, snapshot_file, snapshot_interval):
 
+        if X is None:
+            raise CatBoostError("X must not be None")
+
+        if y is None and not isinstance(X, Pool):
```
Comment (kizill) [inline catboost/python-package/catboost/core.py]: X could also be a string with dataset file path

Comment (mfcabrera) [reply catboost/python-package/catboost/core.py]: @kizill thanks, I saw that but on the documentation of `y` it is not described:
``` y : list or numpy.array or pandas.DataFrame or pandas.Series, optional (default=None)
            Labels, 1 dimensional array like.
            Use only if X is not catboost.Pool.```

If I read this, I believe that  whenever y is None then X should be an instance of Pool. I think the doc shoudl say that. Should I update the docstring as well as the `isinstance` condition?

Comment (mfcabrera) [reply catboost/python-package/catboost/core.py]: @kizill I have added the additional condition.

**Comment by kizill** [comment]
@arcadia-devtools Ship it!

**Comment by arcadia-devtools** [comment]
@kizill, internal review request updated: [808715](https://a.yandex-team.ru/review/808715)

**Comment by arcadia-devtools** [comment]
Check [808715](https://a.yandex-team.ru/review/808715) passed

**Pull request merged**